ee/clickhouse/queries/actor_base_query.py

```diff
@@ -0,0 +1,106 @@
+import uuid
+from typing import (
+    Any,
+    Dict,
+    List,
+    Literal,
+    Optional,
+    Tuple,
+    TypedDict,
+    Union,
+)
+
+from django.db.models.query import QuerySet
+
+from ee.clickhouse.client import sync_execute
+from posthog.models import Entity, Filter, Team
+from posthog.models.filters.mixins.utils import cached_property
+from posthog.models.group import Group
+from posthog.models.person import Person
+
+
+class SerializedPerson(TypedDict):
+    type: Literal["person"]
+    id: uuid.UUID
+    created_at: Optional[str]
+    properties: Dict[str, Any]
+    is_identified: Optional[bool]
+    name: str
+    distinct_ids: List[str]
+
+
+class SerializedGroup(TypedDict):
+    type: Literal["group"]
+    group_key: str
+    created_at: Optional[str]
+    properties: Dict[str, Any]
+
+
+SerializedActor = Union[SerializedPerson, SerializedGroup]
+Actor = Union[Person, Group]
+
+
+class ActorBaseQuery:
+    aggregating_by_groups = False
+
+    def __init__(self, team: Team, filter: Filter, entity: Optional[Entity] = None):
+        self.team = team
+        self.entity = entity
+        self.filter = filter
+
+    def actor_query(self) -> Tuple[str, Dict]:
+        """ Implemented by subclasses. Must return list of uuids. Can be group uuids (group_key) or person uuids """
+        raise NotImplementedError()
+
+    @cached_property
+    def is_aggregating_by_groups(self) -> bool:
+        if self.entity and self.entity.math == "unique_group":
```

Comment:
nit: `return self.entity is not None and self.entity.math == "unique_group"`

Reply:
missed this but added it in the funnel PR